### PR TITLE
Remove requirement for specifying a RID in desktop exe projects

### LINKED ISSUE #396: Remove requirement of setting RuntimeIdentifier in project file for .NET Framework Exe projects
Right now we require that .NET Framework Exe projects [set a `RuntimeIdentifier` property](https://github.com/dotnet/sdk/blob/0908e3556df08d9ba308291b4d9fe4fb2154c0fc/src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets#L49).  This is because ASP.NET Core projects depend on the native library in the [libuv package](https://github.com/aspnet/libuv-package).  In .NET Core we have a way of deploying both x86 and x64 versions and loading the correct one at runtime, but we don't have this for .NET Framework.  That means we effectively don't support AnyCPU for projects that depend on Libuv (or other packages authored in the same way).

At a minimum, we should change it so that the RuntimeIdentifier is only required when there is a package in the dependency graph that has architecture-specific assets.

We should also try to get rid of the requirement of setting`RuntimeIdentifier` in web projects.  This could include setting it to a default in the Web SDK targets when targeting .NET Framework, or updating the Libuv and Kestrel packages to deploy both architectures and load the right native DLL at runtime.

### PR DESCRIPTION
**Pull request by nguerrera**
**Customer scenario**
* Cannot build AnyCPU desktop app (common case) using SDK/new project system without applying a cumbersome and hard-to-understand workaround.
* Cannot build desktop app with native nuget dependencies without making a RID choice.

The approach taken is scoped to address these two common cases:

1. Building AnyCPU desktop application
2. Building win7+ desktop applications with native dependencies

Both of the above will happen automatically and F5 successfully without any mention of RuntimeIdentifier or PlatformTarget in the project. Other scenarios will still require adjusting the RID and PlatformTarget to resolve the native dependencies correctly and stamp the resulting binary correctly.

Follow-ups to improve the experience outside these cases:

* https://github.com/dotnet/sdk/issues/840
* https://github.com/dotnet/sdk/issues/843

**Bugs this fixes:**
Fix #396 and its myriad of duplicates/consequences.

**Workarounds, if any**
Specify dummy runtime identifier for AnyCPU app, and explicitly let PlatformTarget to AnyCPU.
Always pick a RID whenever taking native nuget dependencies from a desktop app.

**Risk**
Low. We address two common cases with new defaults. Specifying PlatformTarget and RuntimeIdentifier explicitly shuts off these defaults and gives the user the same control they had before. We simply make it so that this control needn't be used for the most common cases.

**Performance impact**
Minimal. A few extra checks are needed in msbuild evaluation.

**Is this a regression from a previous update?**
No.

**Root cause analysis**
N/A since we are amending a deliberate design decision in response to significant customer and partner feedback.

**How was the bug found**
Customer and partner reported.

@srivatsn @dsplaisted @eerhardt @dotnet/project-system

### CODE REVIEW

File: TestAssets/TestProjects/DesktopMinusRid/Program.cs
Comment (dsplaisted) [inline TestAssets/TestProjects/DesktopMinusRid/Program.cs]: This seems like the test might eventually break if the Libuv and Kestrel packages get updated to support AnyCPU .NET Framework deployments by deploying both x86 and x64 versions of the native library and Kestrel dynamically chooses which one to load.

I'm not suggesting you change this now, just pointing it out.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: Missing space: `thanwin7`

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: Should we remove this resource (`RuntimeIdentifierMustBeSetForNETFramework`)?  Or is that even something we can do?

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: How likely is it that there could be something else that consumes the `PlatformTarget` property between when we set it to x86 or x64 based on the defaulted RuntimeIdentifier and when we set it back to AnyCPU because we didn't resolve any native assets?

File: TestAssets/TestProjects/DesktopMinusRid/Program.cs
Comment (dsplaisted) [inline TestAssets/TestProjects/DesktopMinusRid/Program.cs]: I'd suggest `typeof(Program).Assembly.Location`.  I don't know that it matters at all, but `Assembly.GetExecutingAssembly()` is one of the methods that we originally removed from .NET Core because it doesn't play well with optimizations we might want to do (inlining across assembly boundaries).

Comment (dsplaisted) [inline TestAssets/TestProjects/DesktopMinusRid/Program.cs]: Naming suggestion: The name `GetProcessorArchitecture` makes me think it will return the architecture of the physical CPU or the OS.  Something like `GetCurrentAssemblyProcessorArchitecture` would be clearer to me, if more verbose.

**Review by dsplaisted — APPROVED** [review]
This looks good.  Congrats on figuring out a reasonable way of improving the user experience here.

**Comment by dsplaisted** [comment]
@mlorbetske @vijayrkn Once this is merged, I believe you should be able to remove the `RuntimeIdentifier` property from the web templates (for example [here](https://github.com/aspnet/Templates/blob/c7e916284e2a2ce4b85628d4d05eaeadfade4e87/src/BaseTemplates/EmptyWeb/EmptyWeb.csproj#L5).  You don't necessarily have to do it for RTM if it doesn't meet the bar though.

@piotrpMSFT @livarcocc Same thing for `dotnet migrate` - If you have logic that sets a `RuntimeIdentifier` property when targeting .NET Framework because we didn't have the same default in the .NET SDK as project.json did, you can probably remove that logic once this is merged.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: Hmm, I'm reviewing the uses of PlatformTarget in msbuild /pp output. There's quite a bit more than I'd hoped.

File: TestAssets/TestProjects/DesktopMinusRid/Program.cs
Comment (nguerrera) [inline TestAssets/TestProjects/DesktopMinusRid/Program.cs]: I'll make the change to avoid any question, but Assembly.GetExecutingAssembly() doesn't suffer from these problems intrinsically. Assembly.GetCallingAssembly is the evil one. GetExecutingAssembly was guilty only by association and that we never bothered to make it intrinsic with identical code gen to typeof(CurrentType).Assembly.Location.

File: TestAssets/TestProjects/DesktopMinusRid/Program.cs
Comment (nguerrera) [inline TestAssets/TestProjects/DesktopMinusRid/Program.cs]: That would only happen if my libuv package ref was upgraded, which is unexpected and I'd be happy to have tests fail in that case.

**Comment by mlorbetske** [comment]
+ @phenning as fyi

**Comment by livarcocc** [comment]
Thanks for the heads up @nguerrera. I have filled https://github.com/dotnet/cli/issues/5668 to track this on the CLI side.

**Comment by nguerrera** [comment]
@MattGertz for approval

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: Removed.

**Comment by nguerrera** [comment]
@MattGertz @natidea @srivatsn Approved to merge or pending ship room?

**Comment by srivatsn** [comment]
Approved at shiproom. However @rrelyea wanted to understand if the RID that we infer is similar to what's done in NuGet.targets.

**Comment by nguerrera** [comment]
@rrelyea @emgarten Can you point me to what nuget.targets does?

**Comment by emgarten** [comment]
@nguerrera this exists in the build task: https://github.com/NuGet/NuGet.BuildTasks/blob/dev/src/Microsoft.NuGet.Build.Tasks/Microsoft.NuGet.targets#L94-L96

NuGet.targets/restore does not infer RIDs anymore like it did in preview2. It only reads the RuntimeIdentifier and RuntimeIdentifiers properties.

**Comment by nguerrera** [comment]
Hold on merging this.

**Comment by nguerrera** [comment]
>>> Approved at shiproom. However @rrelyea wanted to understand if the RID that we infer is similar to what's done in NuGet.targets.
> @nguerrera this exists in the build task: https://github.com/NuGet/NuGet.BuildTasks/blob/dev/src/Microsoft.NuGet.Build.Tasks/Microsoft.NuGet.targets#L94-L96

OK. The default RIDs (plural) there are not win7+ whereas we pick win7+ for the implicit RID here. This is a pragmatic choice to make popular packages like libuv work on F5. Ultimately, the design would be to pick as inclusive as possible for current OS, but that is tracked by https://github.com/dotnet/sdk/issues/840

There is no bad interaction between those and this: restore will just use our RID plus all of the RIDs and produce an assets file with a superset of what this change needs.

**Comment by nguerrera** [comment]
(But still making adjustments based on other issues.)

**Review by natidea — CHANGES_REQUESTED** [review]
Waiting for test

**Comment by nguerrera** [comment]
@natidea @srivatsn @dsplaisted I've got all the changes discussed made and test coverage added to handle it. 

411f959 is unfortunate workaround for case where the net4x app is one target of many. In that case, nuget restore doesn't see the inferred RID of inner build. The fix is a hammer to include win7-x86 and win7-x64 RIDs (plural) in all multi-targeting builds on Windows.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs]: Should we be testing the case where `AppendRuntimeIdentifierToOutputPath` is not specified in the project, as that will be the common case?

**Review by natidea — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: Can you explain this change?  I haven't entirely grokked the difference between DependsOnTargets and AfterTargets, but I think this means that this target could run before ResolvePackageDependenciesForBuild if something else depended on this target.

File: src/Tasks/Microsoft.NET.Build.Tasks/buildCrossTargeting/Microsoft.NET.Sdk.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/buildCrossTargeting/Microsoft.NET.Sdk.targets]: Could we check to see whether `RuntimeIdentifiers` contains a `net*` target framework?  Even just checking if the string contains `net` would allow us to avoid this in most cases where we don't need it.

**Review by dsplaisted — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/buildCrossTargeting/Microsoft.NET.Sdk.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/buildCrossTargeting/Microsoft.NET.Sdk.targets]: It will contain 'net' in most cases: netstandard netcoreapp. I don't want to add regex or other parsing complexity here.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: I am trying to make sure that this is done as soon as we know the right PlatformTarget. Somebody could indeed force it to be sooner.

**Comment by srivatsn** [comment]
Is this ready to merge now?

**Comment by nguerrera** [comment]
Validating Daniel's question about change to AfterTargets and saw that there's an issue with the final change for the multi-targetng case. It won't kick in in full msbuild/VS because of the other issue where Microsoft.Nuget.targets is imported. The satellite test that's covering that is disabled on full msbuild right now.

Working on it.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: I moved it to run soon after the assets are resolved.

**Comment by nguerrera** [comment]
Jenkins is down, but validated locally that d1c2fd4 closes desktop msbuild test gap, and 	5046d92 turns it green. @dsplaisted, @srivatsn please review those changes.

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.RuntimeIdentifierInference.targets]: I'm not really concerned about someone trying to run this target themselves at some other time and I don't want to reset testing for this.

**Comment by nguerrera** [comment]
Do not merge.

The change to deal with multi-targeting case has interactions with #870 and #828 that need to be discussed.

**Comment by natidea** [comment]
@dotnet-bot test this please

**Comment by nguerrera** [comment]
@dsplaisted @natidea @srivatsn I've now fixed the mulitargeting issue in a much more general way that:

* also fixes #871 
* does not regress the lessening of downgrade warnings in #870 
* does not interact with the RID handling of #828 

Please review efbcaca, which does this and is squashed delta since sign-offs. I've also asked @emgarten to review.

I am also going to add some unit tests to check GetAllRuntimeIdentifiers directly and not just in the end-to-end scenario this PR sets out to address.

**Review by srivatsn — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/buildCrossTargeting/Microsoft.NET.Sdk.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/buildCrossTargeting/Microsoft.NET.Sdk.targets]: This condition doesn't seem to be strictly necessary, as this file should only be imported when `TargetFrameworks` is defined, right?

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs
Comment (dsplaisted) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExe.cs]: Should this be checking that the output is built using the correct RID, as is the case with It_handles_native_depdencies_and_platform_target?

**Review by dsplaisted — APPROVED** [review]

**Comment by nguerrera** [comment]
@emgarten approved here https://github.com/dotnet/sdk/issues/871#issuecomment-280185364 

He noted offline that rids are technically case insenstive, so I'll switch to DistinctWithCase().

I am still adding GetAllRuntimeIdentifiers unit tests and I'll add the coverage @dsplaisted requested above. I have to run to an appointment so I'll merge this later tonight once the extra tests are in.

**Comment by nguerrera** [comment]
Abandoning DistinctWithCase because MSBuild is already deduping case elsewhere. I think we have bigger problems if real RIDs ever differ only by case (e.g. path conflicts, endless confusion), so I think we can ignore it.

**Pull request merged**
